### PR TITLE
#184 Adding zpu-go-client

### PR DESCRIPTION
**Pull request by gurleen-gks**

### CODE REVIEW

File: clients/go/zpu/scripts/zpu_set_ownership.pl [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/usr/local/bin/perl -w
```
Comment (havetisyan) [inline clients/go/zpu/scripts/zpu_set_ownership.pl]: this file should be not be included

File: clients/go/zpu/Makefile [hunk 2]
```diff
@@ -0,0 +1,90 @@
+#
+# Makefile to build ZPU Service
+# Prerequisite: Go development environment
+#
+# Copyright 2017 Yahoo Inc.
```
Comment (havetisyan) [inline clients/go/zpu/Makefile]: Company name: Yahoo Holdings, Inc.

File: clients/go/zpu/cmd/tools/main.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package main
```
Comment (havetisyan) [inline clients/go/zpu/cmd/tools/main.go]: Missing license information:

// Copyright 2017 Yahoo Holdings, Inc.
// Licensed under the terms of the Apache version 2.0 license. See LICENSE file for terms.

File: clients/go/zpu/cmd/tools/main.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+import (
+	"flag"
+	"log"
+
```
Comment (havetisyan) [inline clients/go/zpu/cmd/tools/main.go]: let's remove the new lines so from the standard package blocks so they're properly sorted.

File: clients/go/zpu/devel/testing_utilities.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package devel
```
Comment (havetisyan) [inline clients/go/zpu/devel/testing_utilities.go]: missing license details

File: clients/go/zpu/pom.xml [hunk 6]
```diff
@@ -0,0 +1,70 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright 2016 Yahoo Inc.
```
Comment (havetisyan) [inline clients/go/zpu/pom.xml]: 2017 Yahoo Holdings, Inc.

File: clients/go/zpu/test_data/data.go [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package test_data
```
Comment (havetisyan) [inline clients/go/zpu/test_data/data.go]: missing license info

File: clients/go/zpu/test_data/data_domain1.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package test_data
```
Comment (havetisyan) [inline clients/go/zpu/test_data/data_domain1.go]: missing license info

File: clients/go/zpu/test_data/data_domain2.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package test_data
```
Comment (havetisyan) [inline clients/go/zpu/test_data/data_domain2.go]: missing license

File: clients/go/zpu/util/canonicalstring.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
+package util
```
Comment (havetisyan) [inline clients/go/zpu/util/canonicalstring.go]: missing license

File: clients/go/zpu/util/canonicalstring.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+import (
+	"encoding/json"
+	"reflect"
+
```
Comment (havetisyan) [inline clients/go/zpu/util/canonicalstring.go]: remove newline so the standard packages can be sorted in one block

File: clients/go/zpu/util/canonicalstring_test.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
+package util
```
Comment (havetisyan) [inline clients/go/zpu/util/canonicalstring_test.go]: missing license

File: clients/go/zpu/util/canonicalstring_test.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/yahoo/athenz/clients/go/zts"
+
```
Comment (havetisyan) [inline clients/go/zpu/util/canonicalstring_test.go]: remove new lines

File: clients/go/zpu/util/util.go [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package util
```
Comment (havetisyan) [inline clients/go/zpu/util/util.go]: missing license

File: clients/go/zpu/util/util_test.go [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package util
```
Comment (havetisyan) [inline clients/go/zpu/util/util_test.go]: missing license

File: clients/go/zpu/zpu_client.go [hunk 16]
```diff
@@ -0,0 +1,384 @@
+package zpu
```
Comment (havetisyan) [inline clients/go/zpu/zpu_client.go]: missing license

File: clients/go/zpu/zpu_client.go [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+	"os"
+	"strings"
+
+	"github.com/ardielle/ardielle-go/rdl"
+	"github.com/yahoo/athenz/clients/go/zms"
+	"github.com/yahoo/athenz/clients/go/zpu/util"
+	"github.com/yahoo/athenz/clients/go/zts"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+
+	"errors"
```
Comment (havetisyan) [inline clients/go/zpu/zpu_client.go]: this block should be combined with the first block

File: clients/go/zpu/zpu_client_test.go [hunk 18]
```diff
@@ -0,0 +1,371 @@
+package zpu
```
Comment (havetisyan) [inline clients/go/zpu/zpu_client_test.go]: missing license

File: clients/go/zpu/zpu_client_test.go [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/yahoo/athenz/clients/go/zms"
+	"github.com/yahoo/athenz/clients/go/zpu/devel"
+	"github.com/yahoo/athenz/clients/go/zpu/test_data"
+	"github.com/yahoo/athenz/clients/go/zpu/util"
+	"github.com/yahoo/athenz/clients/go/zts"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+
+	"io/ioutil"
```
Comment (havetisyan) [inline clients/go/zpu/zpu_client_test.go]: please combine all standard packages in one block

File: clients/go/zpu/zpu_config.go [hunk 20]
```diff
@@ -0,0 +1,199 @@
+package zpu
```
Comment (havetisyan) [inline clients/go/zpu/zpu_config.go]: missing license

File: clients/go/zpu/zpu_config.go [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
```
Comment (havetisyan) [inline clients/go/zpu/zpu_config.go]: please combine standard blocks

File: clients/go/zpu/zpu_config_test.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
+package zpu
```
Comment (havetisyan) [inline clients/go/zpu/zpu_config_test.go]: missing license

File: clients/go/zpu/zpu_config_test.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+
+import (
+	"testing"
+
```
Comment (havetisyan) [inline clients/go/zpu/zpu_config_test.go]: remove the new line

**Comment by gurleen-gks** [comment]
All the changes are done

**Comment by havetisyan** [comment]
:+1:

**Pull request merged**